File: trunk/EngineUAB/Code/Videogame/Characters/Enemies/Rabbit/States/RabbitAttackState.cpp
#include "RabbitAttackState.h"
#include "Utils\BoostRandomHelper.h"

#include "Characters\StatesDefs.h"
#include "Characters\Enemies\Rabbit\Rabbit.h"
#include "Characters\Character.h"

#include "RabbitPursuitState.h"
#include "RabbitPreparedToAttackState.h"
#include "RabbitTiredState.h"
#include "RabbitDefenseState.h"
#include "RabbitStillAttackState.h"
#include "RabbitRunAttackState.h"

#include "RabbitHitAnimationState.h"
#include "RabbitIdle2AnimationState.h"

#include "Steering Behaviors\SteeringEntity.h"
#include "Steering Behaviors\SteeringBehaviors.h"


#if defined(_DEBUG)
	#include "Memory\MemLeaks.h"
#endif


// -----------------------------------------
//		  CONSTRUCTORS / DESTRUCTOR
// -----------------------------------------
CRabbitAttackState::CRabbitAttackState( void )
	: CState		("CRabbitAttackState")
	, m_ActionTime	( CActionStateCallback( 1.f, 2.f ) )
	, m_pRabbit		( NULL )
{
}

CRabbitAttackState::CRabbitAttackState( const std::string &_Name )
	: CState		(_Name)
	, m_ActionTime	( CActionStateCallback( 1.f, 2.f ) )
	, m_pRabbit		( NULL )
{
}


CRabbitAttackState::~CRabbitAttackState(void)
{
	m_pRabbit = NULL;
}


// -----------------------------------------
//				MAIN METHODS
// -----------------------------------------
void CRabbitAttackState::OnEnter( CCharacter* _Character )
{
	if (!m_pRabbit) 
	{
		m_pRabbit = dynamic_cast<CRabbit*> (_Character);
	}

	// Permite saber si el caracter ha realizado los golpes para cansarse. Ponemos a 2 el hits_done para obligar a que cuando entra en este estado haga una minima pausa antes de atacar
	// _CCharacter.hits_done = 0
	m_pRabbit->SetTotalHitsToBeTired(2);
	m_pRabbit->GetBehaviors()->PursuitOff();
	m_pRabbit->GetBehaviors()->SeekOff();
	/*m_pRabbit->GetBehaviors()->SeparationOn();
	m_pRabbit->GetBehaviors()->CollisionAvoidanceOn();
	m_pRabbit->GetBehaviors()->ObstacleWallAvoidanceOn();*/

}

void CRabbitAttackState::Execute( CCharacter* _Character, float _ElapsedTime )
{
	if (!m_pRabbit) 
	{
		m_pRabbit = dynamic_cast<CRabbit*> (_Character);
	}

	if ( m_pRabbit->IsPlayerAtacable() ) 
	{
		if  ( m_pRabbit->IsFatigued() )
		{
			// print_logger (1, "fatigat després de fer n cops seguits...")
			m_pRabbit->GetLogicFSM()->ChangeState(m_pRabbit->GetTiredState());
		}
		else if ( m_pRabbit->GetReceivedHitsXMinut() == m_pRabbit->GetTotalReceivedHitsXMinut() ) 
		{
			// print_logger (1, "hits x minut rebuts i per tant bloquejaré...")
			m_pRabbit->GetLogicFSM()->ChangeState(m_pRabbit->GetDefenseState());
		}
		else 
		{
			std::string l_ActiveActionState = GetRandomAttackName();
			// print_logger ( 0, "Attack Random Sel·leccionat "..l_ActiveActionState ) 
				
			if ( l_ActiveActionState == STILL_ATTACK_STATE ) 
			{
				m_pRabbit->GetLogicFSM()->ChangeState(m_pRabbit->GetStillAttackState());
			}	
			else if ( l_ActiveActionState == RUN_ATTACK_STATE ) 
			{
				m_pRabbit->GetLogicFSM()->ChangeState(m_pRabbit->GetRunAttackState());
			}	
			else if ( l_ActiveActionState == DEFENSE_STATE ) 
			{
				m_pRabbit->GetLogicFSM()->ChangeState(m_pRabbit->GetDefenseState());
			}		
			// else if ( l_ActiveActionState == "jump" ) then
				// _CCharacter.logic_fsm:change_state(_CCharacter.jump_state)
			else if ( l_ActiveActionState == "go_in_to_fustrum" ) 
			{
				float l_Angle = 22.f;		// 22,5 graus de fustrum
				m_pRabbit->GoInToFrustrum(l_Angle, _ElapsedTime);
			}
		} 	// End fatigue
	}	
	// Si el player NO es atacable lo volvemos a preparar o a perseguir
	else 
	{
		m_pRabbit->GetLogicFSM()->ChangeState(m_pRabbit->GetPreparedToAttack());
	}
}

void CRabbitAttackState::OnExit( CCharacter* _Character )
{
	m_pRabbit->GetBehaviors()->SeparationOff();
	m_pRabbit->GetBehaviors()->CollisionAvoidanceOff();
	m_pRabbit->GetBehaviors()->ObstacleWallAvoidanceOff();
}

bool CRabbitAttackState::OnMessage( CCharacter* _Character, const STelegram& _Telegram )
{
	if ( _Telegram.Msg == Msg_Attack ) 
	{
		// _Character->GetLogicFSM()->ChangeState(m_pRabbit->GetHitState());
		m_pRabbit->GetGraphicFSM()->ChangeState(m_pRabbit->GetHitAnimationState());
		return true;
	}

	return false;
}

// --------------------------------------------------------------------------------------------------------
//	GetRandomAttackName : Devuelve si atacamos de lejos o de cerca y su apropiada animación. También
// 						puede devolver si solo se mueve sin atacar para engañar un poco al player
// --------------------------------------------------------------------------------------------------------
std::string CRabbitAttackState::GetRandomAttackName(void)
{
	std::string l_Action;	

	int l_AttackType = BoostRandomHelper::GetInt(1,12);
	if ( l_AttackType == 1 ) 
		l_Action = STILL_ATTACK_STATE;
	else if ( l_AttackType == 2 ) 
		l_Action =  STILL_ATTACK_STATE;
	else if ( l_AttackType == 3 ) 
		l_Action = STILL_ATTACK_STATE;
	else if ( l_AttackType == 4 ) 
		l_Action = RUN_ATTACK_STATE;
	else if ( l_AttackType == 5 ) 
		l_Action = RUN_ATTACK_STATE;
	else if ( l_AttackType == 6 ) 
		l_Action = RUN_ATTACK_STATE;
	else if ( l_AttackType == 7 ) 
		l_Action = DEFENSE_STATE;
	else if ( l_AttackType == 8 ) 
		l_Action =  JUMP_STATE;
			
	// Más probabilidades de ir al fustrum que no atacar
	else if ( l_AttackType == 9 ) 
		l_Action =  "go_in_to_fustrum";
	else if ( l_AttackType == 10 ) 
		l_Action =  "go_in_to_fustrum";
	else if ( l_AttackType == 11 ) 
		l_Action =  "go_in_to_fustrum";
	else if ( l_AttackType == 12 ) 
		l_Action =  "go_in_to_fustrum";

	return l_Action;
}

File: trunk/Technical_Direction/Tools/Viewer/Data/Scripts/Game/Enemies/rabbit_states/rabbit_attack_state.lua
class 'CRabbitAttackState' (CState)
	function CRabbitAttackState:__init(name) 
		CState.__init(self, name)
		-- print_logger(0, "Inicio del estado gestión de ataques del rabbit")
	end

	function CRabbitAttackState:OnEnter(_CCharacter)
		-- print_logger(0, "CRabbitAttackState:Enter")
		
		-- Permite saber el tiempo de la animación
		self.action_animation_time = 0.0
		
		-- Permite saber si el caracter ha realizado los golpes para cansarse. Ponemos a 2 el hits_done para obligar a que cuando entra en este estado haga una minima pausa antes de atacar
		--_CCharacter.hits_done = 0
		_CCharacter.total_hits_done_to_tired = 2
		_CCharacter.behaviors:pursuit_off()
		_CCharacter.behaviors:seek_off()
		_CCharacter.behaviors:separation_on()
		_CCharacter.behaviors:collision_avoidance_on()
		_CCharacter.behaviors:obstacle_wall_avoidance_on()
	end
	
	function CRabbitAttackState:Execute(_CCharacter, _elapsed_time)
		print_logger(0, "CRabbitAttackState:Execute")
		if ( is_player_attackable( _CCharacter, _CCharacter.player ) ) then
		
			if  ( self:is_fatigued(_CCharacter) ) then 
				-- print_logger (1, "fatigat després de fer n cops seguits...")
				_CCharacter.logic_fsm:change_state(_CCharacter.tired_state)
			elseif ( _CCharacter.received_hits_x_minut == total_received_hits_x_minut ) then 
				-- print_logger (1, "hits x minut rebuts i per tant bloquejaré...")
				_CCharacter.logic_fsm:change_state(_CCharacter.defense_state)
			else 
				_CCharacter.active_action_state = self:GetRandomAttackName()
				print_logger ( 0, "Attack Random Sel·leccionat ".._CCharacter.active_action_state ) 
				
				if  ( _CCharacter.active_action_state == "attack_1" ) then
					_CCharacter.logic_fsm:change_state(_CCharacter.still_attack_state)
				
				elseif ( _CCharacter.active_action_state == "attack_2" ) then
					_CCharacter.logic_fsm:change_state(_CCharacter.run_attack_state)
				
				elseif ( _CCharacter.active_action_state == "defense" ) then
					_CCharacter.logic_fsm:change_state(_CCharacter.defense_state)
					
				-- elseif ( _CCharacter.active_action_state == "jump" ) then
					-- _CCharacter.logic_fsm:change_state(_CCharacter.jump_state)
				
				elseif ( _CCharacter.active_action_state == "go_in_to_frustum" ) then
					local l_angle = 22		-- 22,5 graus de fustrum
					go_in_to_frustum(_CCharacter, l_angle, _elapsed_time)
				
				end 
			
			end 	-- end fatigue
			
		-- Si el player NO es atacable lo volvemos a preparar o a perseguir
		else 
			_CCharacter.logic_fsm:change_state(_CCharacter.prepared_to_attack_state)
		end
	end
	
	function CRabbitAttackState:OnExit(_CCharacter)
		-- print_logger(0, "CRabbitAttackState:Exit")
		-- if not ( _CCharacter == nil ) then
			-- num = _CCharacter:get_animation_id(self.active_animation_name)
			-- _CCharacter:get_animation_model():clear_cycle( num, 0.7 )
		-- end
		_CCharacter.behaviors:separation_off()
		_CCharacter.behaviors:collision_avoidance_off()
		_CCharacter.behaviors:obstacle_wall_avoidance_off()
	end
	
	function CRabbitAttackState:OnMessage(_CCharacter, _Msg)
		-- print_logger(0, "CRabbitAttackState:OnMessage-> Missatge rebut: ".._Msg.msg)
		if ( _Msg.msg == msg_attack ) then
			-- Solo me pongo en hit si realmente he finalizado el estado de ataqueç
			-- print_logger ( 1, "ARA REBO!" )
			if ( self.action_animation_time < _CCharacter.animated_model:get_current_animation_duration(self.active_animation_name) ) then 
				_CCharacter.logic_fsm:change_state(_CCharacter.hit_state)
			else
				_CCharacter:rest_life(  _CCharacter.player.properties.strong )
			end
			return true
		end
		return false
	end
	
	function CRabbitAttackState:__Finalize()
	
	end

	-- --------------------------------------------------------------------------------------------------------
	--	GetRandomAttackName : Devuelve si atacamos de lejos o de cerca y su apropiada animación. También
	-- 						puede devolver si solo se mueve sin atacar para engañar un poco al player
	-- --------------------------------------------------------------------------------------------------------
	function CRabbitAttackState:GetRandomAttackName()
		local AttackType = math.random(1, 12) 
		if ( AttackType == 1 ) then
			return "attack_1"
		elseif ( AttackType == 2 ) then
			return "attack_1"
		elseif ( AttackType == 3 ) then
			return "attack_1"
		elseif ( AttackType == 4 ) then
			return "attack_2"
		elseif ( AttackType == 5 ) then
			return "attack_2"
		elseif ( AttackType == 6 ) then
			return "attack_2"
		elseif ( AttackType == 7 ) then
			return "defense"
		elseif ( AttackType == 8 ) then
			return "jump"
			
		-- Más probabilidades de ir al fustrum que no atacar
		elseif ( AttackType == 9 ) then
			return "go_in_to_frustum"
		elseif ( AttackType == 10 ) then
			return "go_in_to_frustum"
		elseif ( AttackType == 11 ) then
			return "go_in_to_frustum"
		elseif ( AttackType == 12 ) then
			return "go_in_to_frustum"
		end
	end 
	
	function CRabbitAttackState:is_fatigued( _CCharacter )
		-- print_logger ( 1, "hits done : ".._CCharacter.hits_done.." and hits let : ".._CCharacter.total_hits_done_to_tired)
		if ( _CCharacter.hits_done == _CCharacter.total_hits_done_to_tired ) then
			_CCharacter.hits_done = 0
			return true
		else 
			return false
		end 
	end 